### PR TITLE
[refactor][schema] replace linkedin/goavro/v2 with hamba/avro/v2

### PR DESCRIPTION
**Pull request by adrianiacobghiula**
### Motivation

Allow direct deserialization in go struct, allow generation of go struct using avro schema <- support offered by hamba/avro
Improved support in hamba avro project, while linkedin has limited support since 2022 

### Modifications

Replaced linkedin/goavro/v2 with hamba/avro/v2

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change is already covered by existing tests, such as *(please describe tests)*.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API: (no)
  - The schema: ( no )
  - The default values of configurations: (no)
  - The wire protocol: (no)

### Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable / docs / GoDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by adrianiacobghiula** [comment]
Starting from #1191

File: README.md [hunk 1]
```diff
@@ -38,7 +38,7 @@ CGo-based library.
 
 ## Requirements
 
-- Go 1.18+
+- Go 1.20+
```
Comment (crossoverJie) [inline README.md:41]: This is a breaking change, maybe we need more discussion.

Comment (adrianiacobghiula) [reply README.md:41]: this is from the required upgrade of github.com/klauspost/compress v1.17.9

Comment (nodece) [reply README.md:41]: Usually, we want to be compatible with lower versions of Golang, but some libraries force us to upgrade to higher versions.
 
Using 1.20 is in accordance with Golang's release policy: https://go.dev/doc/devel/release

We will release 0.13.0, and I think we can make this change.

File: pulsar/schema.go [hunk 2]
```diff
@@ -125,20 +126,19 @@ func NewSchema(schemaType SchemaType, schemaData []byte, properties map[string]s
 }
 
 type AvroCodec struct {
-	Codec *goavro.Codec
+	Codec avro.Schema
 }
 
-func NewSchemaDefinition(schema *goavro.Codec) *AvroCodec {
-	schemaDef := &AvroCodec{
+func NewSchemaDefinition(schema avro.Schema) *AvroCodec {
```
Comment (nodece) [inline pulsar/schema.go:132]: I worry that this method will be called on the outside(users).

Comment (adrianiacobghiula) [reply pulsar/schema.go:132]: Should this function not be exported ?

Comment (nodece) [reply pulsar/schema.go:132]: This parameter has been changed to `avro.Schema`.

File: README.md (hunk 1 above)
Comment (RobertIndie) [reply README.md:41]: Yes. According to the go release policy:

> Each major Go release is supported until there are two newer major releases.

The currently supported go version is 1.22 and 1.21. We are OK to upgrade the lowest supported go version here.

File: pulsar/schema.go [hunk 3]
```diff
@@ -124,21 +125,10 @@ func NewSchema(schemaType SchemaType, schemaData []byte, properties map[string]s
 	return
 }
 
-type AvroCodec struct {
-	Codec *goavro.Codec
-}
-
-func NewSchemaDefinition(schema *goavro.Codec) *AvroCodec {
```
Comment (RobertIndie) [inline pulsar/schema.go:131]: This is a breaking change. Could we avoid it?

Is it possible to allow users to continue using goavro?

Comment (adrianiacobghiula) [reply pulsar/schema.go:131]: we could, I guess :) -> but that would mean that there is dependency on both linkedin/goavro & hamba/avro 
using hamba internally and just have goavro.

personally I see no benefit of a function that does ... nothing

Comment (adrianiacobghiula) [reply pulsar/schema.go:131]: @RobertIndie keeping `NewSchemaDefinition` function & `AvroCodec` struct with added Deprecation

**Comment by nodece** [comment]
Rethinking recent changes,  due to you changing some method parameters, it appears that there is a breaking change, which resulted in me making an incorrect review.

In the old design, the codec was exported in the schema, and which was created by NewSchemaDefinition(should be internal).

Usually, the users use the following methods to create the avro schema:
```go
NewAvroSchema()
NewAvroSchemaWithValidation()
```

So that we can remove NewSchemaDefinition and AvroCodec.

The next version is the minor version, we can make this changes.

BTW, newXXSchema method returns an implement, not an interface, we also need to improve that in the future.

Is this a fair reason?

**Comment by adrianiacobghiula** [comment]
@nodece fair & done

**Review by nodece — APPROVED** [review]

File: pulsar/schema.go [hunk 4]
```diff
@@ -124,21 +124,10 @@ func NewSchema(schemaType SchemaType, schemaData []byte, properties map[string]s
...
```
Comment (RobertIndie) [inline pulsar/schema.go:131]: Even though the client is still in the development version (0.x.y), it is widely used in production. We should avoid introducing breaking changes in the v0.x version.

We may still need to support goavro here but marked it with deprecation. We can make these breaking changes and remove it in v1.0.

Comment (RobertIndie) [reply pulsar/schema.go:131]: Ok. Now I'm +1.

**Review by RobertIndie — APPROVED** [review]

**Pull request merged**